### PR TITLE
JIT: preliminary version of profile-based inline policy

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Add a new inline policy that can be used when a method has profile data.

It uses the call site frequency to boost profitability. Size and per-call
benefit are currently using the estimates done by the model policy.

Not on by default. Set COMPlus_JitInlinePolicyProfile=1 to enable.
Add testing to weekly experimental runs.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib

**Comment by EgorBo** [comment]
So it fixes https://github.com/dotnet/runtime/issues/41923 (I've just checked) 🎉 

Also, is it expected that it's now able to inline methods with more than `MAX_BASIC_BLOCKS` bbs?
because I've just tested this:
```csharp
string Foo() => "hello";

[MethodImpl(MethodImplOptions.NoInlining)]
bool DoWork()
{
    return Foo() == "hello";
}
```
Currently, inliner gives up here because `string.Equals(string,string)` has 7 basic blocks and ends up with this for `DoWork`:
```asm
G_M53056_IG01:
       sub      rsp, 40
						;; bbWeight=1    PerfScore 0.25
G_M53056_IG02:
       mov      rcx, 0xD1FFAB1E
       mov      rdx, gword ptr [rcx]
       mov      rcx, rdx
       call     System.String:Equals(System.String,System.String):bool
       nop      
						;; bbWeight=1    PerfScore 3.75
G_M53056_IG03:
       add      rsp, 40
       ret
```
Your branch (with profile):
```asm
G_M53056_IG01:              ;; offset=0000H
                                                ;; bbWeight=1    PerfScore 0.00
G_M53056_IG02:              ;; offset=0000H
       B801000000           mov      eax, 1
                                                ;; bbWeight=1    PerfScore 0.25
G_M53056_IG03:              ;; offset=0005H
       C3                   ret
```

File: src/coreclr/src/jit/inlinepolicy.h [hunk 1]
```diff
@@ -305,6 +309,32 @@ class ModelPolicy : public DiscretionaryPolicy
 #endif // defined(DEBUG) || defined(INLINE_DATA)
 };
 
+// ProfilePolicy is an experimental policy that uses the results
+// of data modelling and profile feedback to make estimates.
+
+class ProfilePolicy : public DiscretionaryPolicy
+{
+public:
+    // Construct a ModelPolicy
```
Comment (BruceForstall) [inline src/coreclr/src/jit/inlinepolicy.h:318]: nit: copy-paste comment

File: src/coreclr/src/jit/jitconfigvalues.h [hunk 2]
```diff
@@ -402,6 +402,8 @@ CONFIG_STRING(JitInlineReplayFile, W("JitInlineReplayFile"))
 #endif // defined(DEBUG) || defined(INLINE_DATA)
 
 CONFIG_INTEGER(JitInlinePolicyModel, W("JitInlinePolicyModel"), 0)
+CONFIG_INTEGER(JitInlinePolicyProfile, W("JitInlinePolicyProfile"), 0)
+CONFIG_INTEGER(JitInlinePolicyProfileThreshold, W("JitInlinePolicyProfile"), 40)
```
Comment (BruceForstall) [inline src/coreclr/src/jit/jitconfigvalues.h:406]: String is wrong here: JitInlinePolicyProfile => JitInlinePolicyProfileThreshold

**Review by BruceForstall — APPROVED** [review]
One fix needed in config value string, otherwise LGTM

**Comment by AndyAyersMS** [comment]
> is it expected that it's now able to inline methods with more than `MAX_BASIC_BLOCKS` bbs

Yes. I'm aiming to have very few arbitrary early limits on what might be inlined, to let the profile data and heuristics determine what's best.

We will need to have some limit on the max IL size we'll consider so that we don't spend time analyzing large methods that can't possibly be good inlines. Not sure what the right value for that limit is just yet; I have it set to 1000 bytes here.

File: src/coreclr/src/jit/jitconfigvalues.h (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/jitconfigvalues.h:406]: Thanks for spotting this.

**Comment by AndyAyersMS** [comment]
Failure seems unrelated; none of this code should be enabled by default.

**Pull request merged**